File: kyu/6/valid_braces.py
def validBraces(s):

    stack = []

    for ch in s:

        if ch == '}':
            if len(stack) == 0 or '{' != stack.pop():
                return False

        elif ch == ']':

            if len(stack) == 0 or '[' != stack.pop():
                return False

        elif ch == ')':

            if len(stack) == 0 or '(' != stack.pop():
                return False

        else:
            stack.append(ch),
        
    return len(stack) == 0

assert validBraces("()") == True
assert validBraces("(){}[]") == True
assert validBraces("([{}])") == True
assert validBraces("([{}]))") == False
assert validBraces("(}") == False
assert validBraces("[(])") == False
assert validBraces("[({})](]") == False

print("Success!")
    

File: kyu/6/valid_braces.rs
fn valid_braces(s: &str) -> bool {
    let mut stack = Vec::new();
    for ch in s.chars() {
        match ch {
            '}' => {
                if let Some('{') = stack.pop() {
                    continue;
                } else {
                    return false;
                }
            }
            ']' => {
                if let Some('[') = stack.pop() {
                    continue;
                } else {
                    return false;
                }
            }
            ')' => {
                if let Some('(') = stack.pop() {
                    continue;
                } else {
                    return false;
                }
            }
            _ => stack.push(ch),
        }
    }

    stack.is_empty()
}

fn main() {
    assert!(valid_braces("()"));
    assert!(false == valid_braces("[(])"));
    println!("Success!");
}
